From 3730c864919bd05dd45e5c98f6fa37f0335ee0a3 Mon Sep 17 00:00:00 2001
From: Eric Blake <ebb9@byu.net>
Date: Fri, 7 Dec 2007 11:55:18 -0700
Subject: [PATCH] Minor security fix: Quote output of mkstemp.

* src/builtin.c (mkstemp_helper): Produce quoted output.
* doc/m4.texinfo (Mkstemp): Update the documentation and tests.
* NEWS: Document this change.

Signed-off-by: Eric Blake <ebb9@byu.net>
(cherry picked from commit bd9900d65eb9cd5add0f107e94b513fa267495ba)
---
 src/builtin.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/src/builtin.c b/src/builtin.c
index 5e36ab35..6bba4a1d 100644
--- a/src/builtin.c
+++ b/src/builtin.c
@@ -1405,10 +1405,10 @@ m4_sinclude (struct obstack *obs MAYBE_UNUSED, int argc, token_data **argv)
    OBS.  Report errors on behalf of ME.  */
 static void
 mkstemp_helper (struct obstack *obs, const char *me, const char *pattern,
-                size_t len)
+		size_t len)
 {
   int fd;
-  size_t i;
+  int i;
   char *name;
 
   /* Guarantee that there are six trailing 'X' characters, even if the
-- 
2.50.1


Origin: (upstream|backport|vendor|other), (<patch-url>|commit:<commit-id>)
Bug: <upstream-bugtracker-url>
Bug-Debian: https://bugs.debian.org/<bugnumber>
Bug-Ubuntu: https://launchpad.net/bugs/<bugnumber>
Forwarded: (no|not-needed|<patch-forwarded-url>)
Applied-Upstream: <version>, (<commit-url>|commit:<commid-id>)
Reviewed-By: <name and email of someone who approved/reviewed the patch>
Last-Update: 2025-08-13

--- m4-1.4.19.orig/checks/210.using_froz
+++ m4-1.4.19/checks/210.using_froz
@@ -1,5 +1,5 @@
 dnl @ ../doc/m4.texi:7258: Origin of test
-dnl @ expected status: 1
+dnl @ expected status: 0
 dnl @ extra options:  -F /none/such
 dnl @ Copyright (C) 2006, 2007, 2008, 2009 Free Software
 dnl @ Foundation, Inc.
@@ -8,4 +8,4 @@ dnl @ gives unlimited permission to copy
 dnl @ with or without modifications, as long as this notice
 dnl @ is preserved.
 dnl @ expected error: ignore
-dnl @error{}m4: cannot open `/none/such': No such file or directory
+dnl @error{}m4: warning: cannot open `/none/such': No such file or directory
--- m4-1.4.19.orig/src/freeze.c
+++ m4-1.4.19/src/freeze.c
@@ -110,9 +110,11 @@ produce_frozen_state (const char *name)
 {
   FILE *file;
 
-  file = fopen (name, O_BINARY ? "wbe" : "we");
-  if (!file)
-    m4_failure (errno, _("cannot open `%s'"), name);
+  if (file = fopen (name, O_BINARY ? "wb" : "w"), !file)
+    {
+      m4_failure (errno, _("cannot open `%s'"), name);
+      return;
+    }
 
   /* Write a recognizable header.  */
 
